### PR TITLE
Allow loggers to be passed in

### PR DESCRIPTION
**Pull request by glbrntt**
Motivation:

The community recently discussed how logging should be done for the
`AsyncHTTPClient` project. One of the outcomes was that frameworks
should not log unless they are explicitly passed a logger. This allows
applications to decide whether they care about logging for that
component and have more control over the level at which that component
logs at.

Modifications:

- Client allows for loggers in two places: one for background activity
  (connectivity state management and so on) and another for RPCs
- Server allows a single logger to be specificed
- By default all loggers are no-op loggers
- Added 'none' option for 'CallOptions.RequestIDProvider' for cases
  where the request ID is already present on the logger

Result:

- gRPC does not log by default
- Note: this includes not logging in tests, this will be addressed in a separate PR to
  keep PR size down

### CODE REVIEW

File: Sources/GRPC/ClientConnection.swift [hunk 1]
```diff
@@ -536,7 +549,8 @@ extension ClientConnection {
       connectionKeepalive: ClientConnectionKeepalive = ClientConnectionKeepalive(),
       connectionIdleTimeout: TimeAmount = .minutes(5),
       callStartBehavior: CallStartBehavior = .waitsForConnectivity,
-      httpTargetWindowSize: Int = 65535
+      httpTargetWindowSize: Int = 65535,
+      backgroundActivityLogger: Logger = Logger(label: "io.grpc", factory: { _ in SwiftLogNoOpLogHandler() })
```
Comment (Lukasa) [inline Sources/GRPC/ClientConnection.swift:553]: Is it worth encapsulating this in a `static func` on `Logger` we define internally, just so we have one-and-only-one place to write this code?

File: Sources/GRPC/ClientOptions.swift [hunk 2]
```diff
@@ -60,24 +61,33 @@ public struct CallOptions {
...
+  /// If a `requestIDProvider` exists then a request ID will automatically attached to the logger's
+  /// metadata using the 'grpc-request-id' key.
+  public var logger: Logger
+
   public init(
     customMetadata: HPACKHeaders = HPACKHeaders(),
     timeLimit: TimeLimit = .none,
     messageEncoding: ClientMessageEncoding = .disabled,
-    requestIDProvider: RequestIDProvider = .autogenerated,
+    requestIDProvider: RequestIDProvider = .none,
```
Comment (Lukasa) [inline Sources/GRPC/ClientOptions.swift:74]: What's the reason for this change?

File: Sources/GRPC/PlatformSupport.swift [hunk 3]
```diff
@@ -155,23 +153,22 @@ public enum PlatformSupport {
     networkPreference: NetworkPreference = .best,
     logger: Logger? = nil
```
Comment (Lukasa) [inline Sources/GRPC/PlatformSupport.swift:154]: Why are we defaulting the loggers in this file to `nil` as opposed to the no-op logger?

File: Sources/GRPC/ClientConnection.swift (hunk 1 above)
Comment (glbrntt) [reply Sources/GRPC/ClientConnection.swift:553]: It has to be `public` as it's used in a `public` signature. I didn't think making it underscored-internal was worth it in this case.

Comment (Lukasa) [reply Sources/GRPC/ClientConnection.swift:553]: Ok, fair enough. :smile:

File: Sources/GRPC/PlatformSupport.swift (hunk 3 above)
Comment (glbrntt) [reply Sources/GRPC/PlatformSupport.swift:154]: I don't think there's a good reason to use an optional logger here; I just didn't think to change it.

File: Sources/GRPC/ClientOptions.swift (hunk 2 above)
Comment (glbrntt) [reply Sources/GRPC/ClientOptions.swift:74]: It seemed more natural to opt-in to having request IDs generated rather than opting out. Not sure how much I agree with that thinking now. WDYT?

Comment (Lukasa) [reply Sources/GRPC/ClientOptions.swift:74]: I'm inclined to want to go to "fall into a pit of success": if users don't express a preference, giving them nice UUIDs makes their lives easier, and also our own during bug triage.
